compiler/rustc_ty_utils/src/layout/invariant.rs

```diff
@@ -74,21 +84,19 @@ pub(super) fn sanity_check_layout<'tcx>(cx: &LayoutCx<'tcx>, layout: &TyAndLayou
                 Abi::Uninhabited | Abi::Aggregate { .. },
                 "ABI unexpectedly missing alignment and/or size in {layout:#?}"
             );
-            return;
+            return Ok(());
         };
-        assert_eq!(
-            layout.layout.align().abi,
-            align,
-            "alignment mismatch between ABI and layout in {layout:#?}"
-        );
+        if layout.layout.align().abi != align {
+            return Err(LayoutCalculatorError::ReprConflict);
```

Comment:
Making the `layout_of` query return different results depending on `cfg!(debug_assertions)` seems like a debugging nightmare. Ideally, we should bail before the invalid layout is ever created -- maybe checking for `repr(packed)` in `layout_of_enum` is enough?

Reply:
Yes, but we already "return" an ICE based on `cfg!(debug_assertions)`, which I didn't realize was the reason for the ICE until I found the conditional.

Reply:
This code in practice is not really part of `rustc_ty_utils`, it is `rustc_abi` code that got lost somewhere, and my intent was to not immediately refactor everything before I moved it in there.

Reply:
Like, the problem inheres in the fact that we have "sanity checks" to begin with.

First, I've read that code. I can't call it sane with a straight face.

Second, the entire purpose of a "sanity check" is to attempt to do post-hoc checking of layout invariants, after we've created these invalid layouts. We _do_ have something for that, but it's called a `#[test]`.

Reply:
> Yes, but we already "return" an ICE based on `cfg!(debug_assertions)`

Yeah that's how debug asserts work? If `debug_assertions` are enabled they sometimes "return" a panic rather than the wrong thing, but here you're making it so that if `debug_assertions` are enabled, we're suddenly starting to return the correct thing instead, which seems confusing to me.

> Like, the problem inheres in the fact that we have "sanity checks" to begin with.

I don't see the problem with sanity checks here or the advantage of moving this code to a `#[test]`. When the sanity check is enabled (by enabling debug assertions) it's like we're running the test for every layout that's computed by the compiler and that's how we're finding the bugs like the one you're fixing here in the first place. If we just `#[test]` some list of predefined layouts, then we would have never found this bug.

Reply:
Like, running the sanity check should always be optional (like a test), and never affect the end result of `layout_of` modulo internal errors being reported (via ICEs).

Reply:
> I don't see the problem with sanity checks here or the advantage of moving this code to a `#[test]`. When the sanity check is enabled (by enabling debug assertions) it's like we're running the test for every layout that's computed by the compiler and that's how we're finding the bugs like the one you're fixing here in the first place. If we just `#[test]` some list of predefined layouts, then we would have never found this bug.

For the most part, we _do_ test it on predefined layouts, though: the ones we actually include in test-cases and run with debug assertions enabled on. If the asserts worked to catch all these cases in general, we would not have merged #123367 or #128612 in the condition they were merged in.

...well, actually it seems #128612 had bors-related errors surrounding its merging so I guess it would've. 😅

Reply:
i.e. The suite of check-fail tests doesn't really drive into the layout computing code without `-Zvalidate-mir` forcing it to, which is, er, why `-Zvalidate-mir` exists now.

Reply:
As far as tests go, real property testing at the source of the problem can actually exhaust a possibility space much faster than randomly mutating Rust source programs and then calling rustc on them with `-Zvalidate-mir`.

Reply:
anyways what I really mean is that having a sanity check post-hoc, and returning errors to begin with, means we've already conceded that layout computation will compute erroneous layouts, so the critique of "we shouldn't have the invariant check return different results if the layout is invalid, we should fix it at the source" is valid, but imo applies in a more general sense to having _any_ code that checks invariants "downstream" of `rustc_abi`.

Reply:
These are all great points for moving the invariant check into `rustc_abi` (and creating a dedicated test suite), but forcing to run the check on every layout after it has been constructed to convert from a `Ok(layout)` into an `Err(something)` still seems post-hoc to me.

For this specific issue I believe the proper fix at the source would be to just check for `repr(packed)` before computing the layout of an enum and returning an error immediately rather than continue and then reject the computed layout later.